### PR TITLE
dev/core#1378 Scheduled reminders: do not email if do_not_email or on_hold

### PR DESCRIPTION
**Pull request by agh1**
*Note: this PR used to just be a test, but I figured out how to easily fix it*

Overview
----------------------------------------
This suppresses scheduled reminder emails for contacts that have "Do not email" or who have their primary email on hold.  This matches the behavior for SMS scheduled reminders

This also adds to the scheduled reminder tests a handful of cases to make sure that people with "Do not email", dead people, and people with their primary email address on hold are not emailed scheduled reminders.

See https://lab.civicrm.org/dev/core/issues/1378

Before
----------------------------------------
People with `do_not_email=1` nevertheless receive emailed scheduled reminders, as do people with emails where `on_hold=1`.

People with `is_deceased=1` do not receive scheduled reminders at all, nor do people with `do_not_sms=1` receive SMS scheduled reminders.

After
----------------------------------------
Emailed scheduled reminders are not sent if the contact has `do_not_email=1`, `is_deceased=1`, or the primary email with `on_hold=1`.  SMS scheduled reminders are not sent if the contact has `do_not_sms=1` or `is_deceased=1`.

Technical Details
----------------------------------------
Nothing that isn't otherwise obvious.  I also added a test case for "Gretchen" who has a primary email address that is not on hold and another email address that is on hold: she should and does receive reminders to her primary email.  I do not account for the reverse situation, where the primary address is on hold but some other address is not.  We can't/shouldn't try to read the contact's mind.

My edit follows the lead of the SMS code by skipping the contact at the level of finding the email address rather than excluding them from the overall list of recipients.  This means that if someone has an email on hold or has Do not email, they might still get the reminder if it's also sent as SMS.

Comments
----------------------------------------
This will also require a fix to the documentation to edit this: https://github.com/civicrm/civicrm-user-guide/blob/master/docs/email/scheduled-reminders.md#scheduled-reminders-and-privacy-options

### CODE REVIEW

**Comment by agh1** [comment]
I just thought I'd add a test to ensure that reminders aren't sent to emails that are on hold.  Lo and behold, the test fails because we *do* send reminders to known bad email addresses.

**Comment by seamuslee001** [comment]
@agh1 style fix needed https://test.civicrm.org/job/CiviCRM-Core-PR/30386/checkstyleResult/new/

**Comment by agh1** [comment]
Thanks @seamuslee001 fixed it

**Comment by eileenmcnaughton** [comment]
This hurts my brain - we had conversations before about Scheduled reminders sending emails when people thought they 'shouldn't' but Dave G said their were transactional not bulk mailings - perhaps that was about respecting is_opt_out - @petednz can you recall?

**Comment by agh1** [comment]
Test failure *is* related :wink:

**Comment by agh1** [comment]
FWIW scheduled reminders *do* respect `do_not_sms` preferences:
https://github.com/civicrm/civicrm-core/blob/b30514b39bf8abda38c98b1a019f1213f0d1bcb5/CRM/Core/BAO/ActionSchedule.php#L678-L683

**Comment by agh1** [comment]
As you can see from the edit to the description and the latest commits, I figured out a fix for this.  After seeing that SMS already works as I expected, I feel reassured that this is a change worth making.

File: CRM/Contact/BAO/Contact.php [hunk 1]
```diff
@@ -2560,18 +2560,26 @@ public static function matchContactOnOpenId($openId, $ctype = NULL) {
...
    *   Contact id.
+   * @param bool $polite
+   *   Whether to only pull an email if it's okay to send to it--that is, if it
+   *   is not on_hold and the contact is not do_not_email.
    *
    * @return string
    *   Email address if present else null
    */
-  public static function getPrimaryEmail($contactID) {
+  public static function getPrimaryEmail($contactID, $polite = FALSE) {
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Contact.php]: I love it

**Comment by eileenmcnaughton** [comment]
This looks good to  me - @jitendrapurohit are you maybe able to review?

**Comment by jitendrapurohit** [comment]
looks like a straight-forward fix. I'll start reviewing tomorrow  👍

**Comment by jitendrapurohit** [comment]
works great and correctly avoids sending reminders to contacts having `on_hold` or `do not email` = 1. But - seems like this could be a problem if on_hold/donotemail is removed from contact just after the sched reminder job is executed?  I see the current code enters a record in civicrm_action_log table so the below case will not work after this PR, i.e -

- Create a contact with do_not_email = 1.
- Configure a sched reminder so that it gets sent to our test contact.
- Execute the reminder job. No email is sent(correct).
- Disable do_not_email on the contact.
- Run the job again. The email is still not sent to the email address(Incorrect?).

This seems to be because, on first execution of the job, civi enters a row in `civicrm_action_log` table with is_error = 1 and message = `Couldn't find recipient's email address`.  The message seems incorrect and shouldn't have inserted in this table?

As there is an existing record with the same values, reminder job does not process them again.

**Comment by agh1** [comment]
I can see that potential use case of "whoops, some of these people have 'on hold' emails and didn't get the reminder--let me unset those and resend", but it seems like an extreme niche case.

Meanwhile, if old reminders that were intended to be sent while you have "do not email" or "on hold" were to go out after you unset that preference, wouldn't that potentially unleash a torrent of reminders from years past whenever someone unsets "on hold" or "do not email"?

**Comment by agh1** [comment]
Also, one goal of this is to ensure email and SMS work in a parallel fashion.  The email code:
https://github.com/civicrm/civicrm-core/blob/83aa0e630490d90fb24f097d86a254bd09bae0ff/CRM/Core/BAO/ActionSchedule.php#L610-L613
is written to mimic the SMS behavior:
https://github.com/civicrm/civicrm-core/blob/83aa0e630490d90fb24f097d86a254bd09bae0ff/CRM/Core/BAO/ActionSchedule.php#L544-L547
and the phone number lookup in the SMS code excludes numbers where someone has "do not SMS":
https://github.com/civicrm/civicrm-core/blob/83aa0e630490d90fb24f097d86a254bd09bae0ff/CRM/Core/BAO/ActionSchedule.php#L679-L683

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit from @agh1 comments it seems like he is suggesting that this should be merged as an improvement & that it does not make the situation you identified worse - is that correct & if so should I merge this now?

**Comment by totten** [comment]
> This hurts my brain - we had conversations before about Scheduled reminders sending emails when people thought they 'shouldn't' but Dave G said their were transactional not bulk mailings - perhaps that was about respecting is_opt_out - @petednz can you recall?

Yeah, this is a bit tricky. I'm not sure if there's anything written about untangling that... taking a stab at verbalizing the problem...

* The *mechanism* of a "scheduled email" can be used for transactional purposes ("We need to know about your dietary needs for the conference's lunch"; "Please bring a laptop to tomorrow's training event") or for marketing purposes ("You may like to know about our next event"; "Here's some cool stuff we're doing"). If someone has `No bulk emails/is_opt_out`, then I'd expect the first emails to through anyway, and the latter emails to be held back.
    * (*Thoughtlet: Each reminder needs a flag/category? Maybe "Transactional"/"Marketing"?*)
* The discussion touches on multiple fields: `do_not_email`, `do_not_sms`, `on_hold`, `is_deceased`. As Andrew points out, scheduled-reminders don't handle `do_not_email` and `do_not_sms` in the same way.
* Both the "Before" and "After" behavior have precedent elsewhere in the system. For example, suppose a contact has `do_not_email=1` and I go into the backend UI to make some records about them...
    * If I make a "Contribution" and enable the checkbox to send a confirmation, then it sends the email (*obeying the admin's choice; ignoring the contact's choice*).
    * If I make a "Participant" and enable the checkbox to send a confirmation, then it skips the email (*obeying the contact's choice; ignoring the admin's choice*).
    * (*Thoughtlet: If the contact has `do_not_email=1`, then backend UI should inform the admin, but the admin should still have the choice of whether to send the msg?*)
* IMHO, reasonable people will look at a "Do not email" checkbox in this system and make different assumptions. Some would read "Do not email" subjectively as an imperative that *staff should not initiate contact over email*, and some would read "Do not email" literally as *a total blackout/prohibition on email*. (There are several things in the grey-area between those interpretations: receipting, password-recovery, giving or asking for information about event attendance, responding to emails from the contact, etc.)

**Comment by agh1** [comment]
> - If I make a "Contribution" and enable the checkbox to send a confirmation, then it sends the email (obeying the admin's choice; ignoring the contact's choice).
> - If I make a "Participant" and enable the checkbox to send a confirmation, then it skips the email (obeying the contact's choice; ignoring the admin's choice).

This is ridiculous--really?  I could see rationales for either, and I like @totten's thoughtlet about the option to override, but it's so classic that we do the same thing in two places and silently do them opposite ways.

Back to reminders, I outlined a few ways that people send emails, and I've cross-referenced them with what someone of the *staff should not initiate contact over email* mindset and someone of the *total blackout/prohibition on email* mindset might think, along with the current behavior.

| Message               | Method          | Don't initiate email | Absolutely no email | Current behavior     |
| --------------------- | --------------- | -------------------- | ------------------- | -------------------- |
| Please join us        | CiviMail        | suppress             | suppress            | suppress             |
| Event venue moved     | "Send an email" | suppress             | suppress            | suppress             |
| Online contribution   | Payment receipt | allow                | allow               | allow                |
| Event is tomorrow     | Sched. reminder | ?                    | suppress            | allow email, not sms |
| Renew your membership | Sched. reminder | ?                    | suppress            | allow email, not sms |
| Thank you for coming  | "Send an email" | suppress             | suppress            | suppress             |
| Thank you for coming  | Sched. reminder | ?                    | suppress            | allow email, not sms |
| Happy birthday        | Sched. reminder | suppress             | suppress            | allow email, not sms |
| Welcome series        | Sched. reminder | suppress             | suppress            | allow email, not sms |
| Update your CPR cert. | Sched. reminder | ?                    | suppress            | allow email, not sms |
| Update your CPR cert. | "Send an email" | suppress             | suppress            | suppress             |

I put in question marks for a few situations where I, a hard-liner, think it would be totally inappropriate to send emails to someone who doesn't want emails, but I want to be generous in case someone thought it would be desirable in those cases to ignore the preference.

Hopefully this grid is helpful for figuring things out.  To me, it's pretty clear that if there is going to be blanket behavior by scheduled reminders, it should be to suppress emails to Do Not Email contacts, but that an option on the reminder to override people's wishes would be a welcome new feature (albeit outside the scope of this PR).

**Comment by petednz** [comment]
I am struggling to recall a specific discussion on this with DG but the scenario it may have covered would be as follows:

a professional organisation (eg an Association of Surgeons) have a 'requirement' to be able to notify members about eg
- the need to renew
- the need to maintain their professional certification.

Let's say the org also sends out bulk mailings, and some folk use the OPT OUT option with the expectation that this will stop all such bulk mailings, but that they will continue to get the above reminders

**Comment by petednz** [comment]
In terms of Totten's thoughtlets (interesting term)
- Each reminder needs a flag/category? Maybe "Transactional"/"Marketing"
- If the contact has do_not_email=1, then backend UI should inform the admin, but the admin should still have the choice of whether to send the msg

I think these both would go a long way to let us have the choices needed to step around these tricky interpretations about what the system should/could do.

**Comment by totten** [comment]
@agh1 The table is a nice idea! Here are some more comments on the "?" items and on some of the others: https://docs.google.com/spreadsheets/d/1RpfF3DRrTYDWFTsE8ebyY3mjQJc9BeXjpFO4GzxGcrQ/edit?usp=sharing (see tab "r2"). I also joined the GL thread.

> (@petednz) a professional organisation (eg an Association of Surgeons) have a 'requirement' to be able to notify members about eg
>
>    the need to renew
>    the need to maintain their professional certification.
>
> Let's say the org also sends out bulk mailings, and some folk use the OPT OUT option with the expectation that this will stop all such bulk mailings, but that they will continue to get the above reminders

Just checking my impression here...  with this PR, the above case still works for those folks, right? It just has a couple assumptions: (1) that optouts have been correctly stored in `is_opt_out` , (2) that the important messages about renewal/certification are specifically sent as scheduled-reminders, and (3) that the less-important messages are specifically sent with CiviMail/Mosaico.

The potential problem scenarios would be cases which break those assumptions, e.g. cases where: (1)  the admins were confused about `is_opt_out`/`do_not_email` (*when configuring a profile/webform, the `is_opt_out`/`do_not_email` distinction is clear as mud*), (2) the important messages are sent through CiviMail/Mosaico, or (3) the less-important messages are sent with Scheduled Reminders.

I guess what we need to figure is if those problem scenarios are real/common or just straw-men?

**Comment by agh1** [comment]
For your questions @totten:

1. People are frequently confused, but the solution shouldn't be to blur the lines further.  If we're clear that `do_not_email` is strict and `is_opt_out` mainly excludes CiviMail--and people experience that `do_not_email` people actually are excluded from almost all emails--that will help make this useful.
2. Yes, lots of important messages are sent through CiviMail.  A classic example is important announcements for an event with more than 50 people.  However, with being responsible email senders, we can't do anything about this besides always offering *both* opt-out and unsubscribe links and being good about mailing list groups.
3. Yes, less-important emails do get sent with scheduled reminders.  One client sends everyone a "happy birthday" email (and commissioned the work to do reminders based on contact date fields in order to do so).  A number of organizations send a welcome series of emails to new members--marketing content that would ordinarily be sent in a CiviMail message except that it's sent relative to the date when someone joins.  It would be ideal to exclude `is_opt_out` people from these or even offer an "unsubscribe group" like when you send to search results in CiviMail.

These are real, common scenarios, but they're not the default, and none of them are served by having scheduled reminders ignore `do_not_email` and `on_hold` across the board.

(One note: my goal with the `$polite` param in `CRM_Contact_BAO_Contact::getPrimaryEmail()` is that different uses could be explicit about whether they're looking to ignore preferences--and maybe it could be set based upon a per-reminder or per-receipt setting like folks have imagined here.)

**Comment by agileware-justin** [comment]
> To me, it's pretty clear that if there is going to be blanket behavior by scheduled reminders, it should be to suppress emails to Do Not Email contacts, but that an option on the reminder to override people's wishes would be a welcome new feature (albeit outside the scope of this PR).

Agreed and let's get this PR merged. Jeepers!

**Comment by agh1** [comment]
@totten the consensus over in GitLab seems that the baseline behavior should be what this PR does, even though we all wish there were features for finer control.  Do you have concerns about this PR?

**Comment by totten** [comment]
@agh1 Alrighty, let's do it!

**Comment by agileware-justin** [comment]
PR to update documentation with this change, https://lab.civicrm.org/documentation/docs/user-en/-/merge_requests/580

**Pull request merged**